### PR TITLE
Implement async blocks

### LINKED ISSUE #4018: Implement async blocks
We don't currently handle async blocks correctly:
```rust
let x = async { 1 }.await;
```
Basically, we just ignore the `async` modifier at the moment. We need to:
 - add the information whether the block is `async` to [`Expr::Block`](https://github.com/rust-analyzer/rust-analyzer/blob/690a0b249eee6537953350be28a0ca11fa96b12a/crates/ra_hir_def/src/expr.rs#L52-L55) and set this during [lowering](https://github.com/rust-analyzer/rust-analyzer/blob/690a0b249eee6537953350be28a0ca11fa96b12a/crates/ra_hir_def/src/body/lower.rs#L467) (this might need some adaptation to ast_src.rs and maybe even the parser to get the async token)
 - adapt [`infer_block`](https://github.com/rust-analyzer/rust-analyzer/blob/690a0b249eee6537953350be28a0ca11fa96b12a/crates/ra_hir_ty/src/infer/expr.rs#L521) to wrap the result in an `impl Future` if the block is async. This requires impl trait support, which we don't have at the moment but should be ready in Chalk soon.

### PR DESCRIPTION
**Pull request by oxalica**
Fix #4018

@flodiebold already gave a generic guide in the issue. Here's some concern about implementation detail:
- Chalk doesn't support generator type yet.
- Adding generator type as a brand new type (ctor) can be complex and need to *re-introduced* builtin impls. (Like how we implement closures before native closure support of chalk, which is already removed in #5401 )
- The output type of async block should be known after type inference of the whole body.
  - We cannot directly get the type from source like return-positon-impl-trait. But we still need to provide trait bounds when chalk asking for `opaque_ty_data`.
  - During the inference, the output type of async block can be temporary unknown and participate the later inference.
    `let a = async { None }; let _: i32 = a.await.unwrap();`

So in this PR, the type of async blocks is inferred as an opaque type parameterized by the `Future::Output` type it should be, like what we do with closure type.
And it really works now.

Well, I still have some questions:
- The bounds `AsyncBlockImplType<T>: Future<Output = T>` is currently generated in `opaque_ty_data`. I'm not sure if we should put this code here.
- Type of async block is now rendered as `impl Future<Output = OutputType>`. Do we need to special display to hint that it's a async block? Note that closure type has its special format, instead of `impl Fn(..) -> ..` or function type.

### CODE REVIEW

File: crates/hir_ty/src/lib.rs [hunk 1]
```diff
@@ -843,6 +848,33 @@ impl Ty {
 
     pub fn impl_trait_bounds(&self, db: &dyn HirDatabase) -> Option<Vec<GenericPredicate>> {
         match self {
+            Ty::Apply(ApplicationTy { ctor: TypeCtor::OpaqueType(opaque_ty_id), parameters }) => {
```
Comment (flodiebold) [inline crates/hir_ty/src/lib.rs:851]: why is this necessary?

File: crates/hir_ty/src/traits/chalk.rs [hunk 2]
```diff
@@ -166,27 +168,86 @@ impl<'a> chalk_solve::RustIrDatabase<Interner> for ChalkContext<'a> {
...
+                    .lang_item(self.krate, "future_trait".into())
+                    .and_then(|item| item.as_trait())
+                    .and_then(|trait_| {
+                        let alias =
+                            self.db.trait_data(trait_).associated_type_by_name(&name![Output])?;
+                        Some((trait_, alias))
+                    })
+                {
+                    // AsyncBlock<T>: Future</* Self */>
+                    // This is required by `fn impls_future` to check if we need to provide `.await` completion.
```
Comment (flodiebold) [inline crates/hir_ty/src/traits/chalk.rs:206]: it's not just required by that, it's just required in general if we want async blocks to implement `Future` ;)

Comment (oxalica) [reply crates/hir_ty/src/traits/chalk.rs]: The statement is removed now.

File: crates/hir_ty/src/lib.rs (hunk 1 above)
Comment (oxalica) [reply crates/hir_ty/src/lib.rs]: This method is used in type walking, which is used for "goto type" hover.
Seems that I did it in a wrong way before.
But now it's fixed and a test is added. It's required to provide trait "Future" in "goto type" hover.

**Comment by flodiebold** [comment]
Thanks!

bors r+

**Comment by matklad** [comment]
Oh wow, just as I took a vacation :D 

Kudos!
